### DISCUSSION TITLE
Edge Dilation for Depth Maps

### DISCUSSION THREAD

**Discussion by WorldofDepth** [ImageMagick/ImageMagick#6648] [Help]
### ImageMagick version

7.0.8-3

### Operating system, version and so on

Mac OS X 10.10.5

### Description

Hi, I am trying to adjust edge dilation in grayscale depth maps. Basically, this means slightly expanding all areas of brighter pixels _on top_ of darker pixels—i.e. with priority according to brightness. (For reference: http://www.yuvsoft.com/stereo-3d-technologies/dilate-depth-map/)

I am getting close to what I want for this example depth map using `-morphology dilate diamond:8`, but the problem is the added edge pixels are significantly brighter than their abutting areas. On close examination, there are edge (anti-aliasing?) artifacts in the original, including some bright pixels, and I think this is the cause. Can anyone suggest a solution, to either correct the artifacts in the original image first, or somehow implement a neighborhood-aware dilation effect? Thank you for any help!

Images: original / with dilation / with dilation close-up
![image](https://github.com/ImageMagick/ImageMagick/assets/77465052/57fcf886-db53-4ffe-8deb-7f032940ddd1)
![image](https://github.com/ImageMagick/ImageMagick/assets/77465052/ef4c28bd-f958-4cbc-8921-b27820b3ffea)
<img width="462" alt="image" src="https://github.com/ImageMagick/ImageMagick/assets/77465052/10daca16-50c0-4543-b1a1-39cc7a563eb5">

**Comment by fmw42** [comment]
I do not see the bright pixels causing your issue.  Can you point them out? In fact, I do not see where the pixels are getting brighter.  The edge pixels should be brighter when doing -morphology dilate, since they will expand into the darker surrounding.  Please clarify your issue.

**Reply by WorldofDepth** [reply]
Thanks for the reply! The brighter pixels are expanding into the surrounding darker areas as desired, but they are brighter than they should be. The difference is a bit hard to see—hopefully it's more visible below. Can you can see the fuzzy brighter border to the light area in the dilated version?

![image](https://github.com/ImageMagick/ImageMagick/assets/77465052/eede8f9a-221c-43c0-b925-f8680c9e44f4)

**Comment by snibgo** [comment]
Areas near the edge of the head are darker than areas near the centre. "`-morphology dilate`" lightens pixels that are close to lighter pixels. That is what it should do. That is what you want where it causes a large change, but not where it causes a small change.

This suggests you could create a mask, so the effect occurs only where there are large changes.

**Reply by WorldofDepth** [reply]
Thanks for the reply. Do you think there's any way to cap the brightness of the lightened pixels, so they don't exceed the brightness of their neighbors?

**Comment by snibgo** [comment]
> Do you think there's any way to cap the brightness of the lightened pixels, so they don't exceed the brightness of their neighbors?

That won't do what you want. You *do* want pixels near edges to lighten and exceed the brightness of their neighbors.

I'm not sure exactly what you want. I understand roughly, but not exactly.

Possibly: You want to change only pixels within a certain distance of an edge. (What distance? What exactly is an "edge"?) On the light side of the edge, no pixels should change. On the dark side, you want the pixels to become lighter, the same colour as the nearest pixel on the light side.

**Reply by WorldofDepth** [reply]
Ah yes, I meant they shouldn't exceed the brightness of their _brightest_ neighbors. Only match it.

Basically, the goal is to expand each area of light pixels, into the surrounding darker pixel areas. But the brightness _range_ and minimums/maximums should not change. Does that suggest any different approaches?

Thank you again for the help!

**Comment by snibgo** [comment]
We can set every pixel to be the darker of (a) the morphology you showed or (b) the lightest pixel within a window around that pixel.

Morphology diamond:8 has a radius of 8 pixels, so we need a window of 17x17 pixels.

Bash script:
```
magick depthMap.png \
  \( -clone 0 -morphology dilate diamond:8 +write d.png \) \
  \( -clone 0 -statistic Maximum 17x17 +write x.png \) \
  -delete 0 -compose Darken -composite \
  x2.png
```
But you might prefer a smaller window, eg 15x15 or 13x13.
